fastjet/internal/delaunay/triangulation.go

```diff
@@ -0,0 +1,663 @@
+// Copyright 2017 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package delaunay
+
+import (
+	"fmt"
+)
+
+// holds necessary information for the
+// delaunay triangulation
+type Delaunay struct {
+	triangles []*Triangle // all triangles created
+	root      *Triangle   // triangle that contains all points. Used to find a triangle
+	maxX      float64
+	minX      float64
+	maxY      float64
+	minY      float64
+}
+
+// creates a delaunay triangulation with the given points
+// all points have to be inside the user defined bounds
+func NewDelaunay(points []*Point, maxX float64, minX float64, maxY float64, minY float64) *Delaunay {
+	// root Triangle is a triangle that contains all points
+	dx := maxX - minX
+	dy := maxY - minY
+	a := NewPoint(minX-8*dx, maxY+10, -1)
+	b := NewPoint(maxX+8*dx, maxY+10, -1)
+	c := NewPoint(minX+dx/2, minY-8*dy, -1)
+	root := NewTriangle(a, b, c)
+	d := &Delaunay{
+		root: root,
+		maxX: maxX,
+		minX: minX,
+		maxY: maxY,
+		minY: minY,
+	}
+	d.triangles = make([]*Triangle, 0)
+	for _, p := range points {
+		d.InsertPoint(p)
+	}
+	return d
+}
+
+func (d *Delaunay) GetTriangles() []*Triangle {
+	// remove triangles that contain the root points
+	rt := make([]*Triangle, 0, len(d.root.A.adjacentTriangles)+len(d.root.B.adjacentTriangles)+len(d.root.C.adjacentTriangles))
+	rt = append(rt, d.root.A.adjacentTriangles[:]...)
+	rt = append(rt, d.root.B.adjacentTriangles[:]...)
+	rt = append(rt, d.root.C.adjacentTriangles[:]...)
+	return finalize(d.triangles, rt[:]...)
+}
+
+func (d *Delaunay) InsertPoint(p *Point) {
+	p.adjacentTriangles = make([]*Triangle, 0)
+	t, onE := findTriangle(d.root, p)
+	if t == nil {
+		// should only happen when user gives wrong max and min values
+		panic(fmt.Errorf("delaunay: no triangle which contains P%s", p))
+	}
+	if onE {
+		d.insertPonE(t, p)
+	} else {
+		d.insertP(t, p)
+	}
+}
+
+func (d *Delaunay) RemovePoint(p *Point) {
+	// if that's the case either an error has occurred or the point is outside or on the root triangle
+	if len(p.adjacentTriangles) < 3 {
+		// should only happen when user makes a mistake
+		panic(fmt.Errorf("delaunay: can't remove point P%s", p))
```

Comment:
`panic(fmt.Errorf("delaunay: can't remove point P%s not enough triangles", p))` or something similar.